serialdevice.go

```diff
@@ -144,22 +167,34 @@ func (dev SerialDevice) QemuParams(config *Config) []string {
 	var qemuParams []string
 
 	deviceParams = append(deviceParams, dev.deviceName(config))
-	if s := dev.Transport.disableModern(config, dev.DisableModern); s != "" {
-		deviceParams = append(deviceParams, s)
-	}
 	deviceParams = append(deviceParams, fmt.Sprintf("id=%s", dev.ID))
-	if dev.Transport.isVirtioPCI(config) && dev.ROMFile != "" {
-		deviceParams = append(deviceParams, fmt.Sprintf("romfile=%s", dev.ROMFile))
-		if dev.Driver == VirtioSerial && dev.MaxPorts != 0 {
-			deviceParams = append(deviceParams, fmt.Sprintf("max_ports=%d", dev.MaxPorts))
+	if dev.Addr != "" {
+		deviceParams = append(deviceParams, fmt.Sprintf("addr=%s", dev.Addr))
+	}
+	if dev.ROMFile != "" {
+		if dev.Driver == PCISerial || dev.Transport.isVirtioPCI(config) {
+			deviceParams = append(deviceParams, fmt.Sprintf("romfile=%s", dev.ROMFile))
 		}
 	}
-
-	if dev.Transport.isVirtioCCW(config) {
-		if config.Knobs.IOMMUPlatform {
-			deviceParams = append(deviceParams, "iommu_platform=on")
+	switch dev.Driver {
+	case VirtioSerial:
+		if s := dev.Transport.disableModern(config, dev.DisableModern); s != "" {
+			deviceParams = append(deviceParams, s)
+		}
+		if dev.Transport.isVirtioPCI(config) && dev.MaxPorts != 0 {
+			deviceParams = append(deviceParams, fmt.Sprintf("max_ports=%d", dev.MaxPorts))
+		}
+		if dev.Transport.isVirtioCCW(config) {
+			if config.Knobs.IOMMUPlatform {
+				deviceParams = append(deviceParams, "iommu_platform=on")
+			}
+			deviceParams = append(deviceParams, fmt.Sprintf("devno=%s", dev.DevNo))
+		}
+	case PCISerial:
+		deviceParams = append(deviceParams, fmt.Sprintf("chardev=%s", dev.ChardevID))
+		if dev.Multifunction {
```

Comment:
It's always worth checking the features of a given driver, VirtioSerial driver string is 'virtio-serial'; to see the options it supports you can run:

```
$ qemu-system-x86_64 -device virtio-serial,?
virtio-serial-pci options:
notify_on_empty=<bool> - on/off
use-started=<bool>
emergency-write=<bool> - on/off
max_ports=<uint32>
x-ignore-backend-features=<bool>
virtio-backend=<child<virtio-serial-device>>
packed=<bool>          - on/off
multifunction=<bool>   - on/off
page-per-vq=<bool>     - on/off
vectors=<uint32>
x-pcie-deverr-init=<bool> - on/off
disable-legacy=<OnOffAuto> - on/off/auto
x-pcie-pm-init=<bool>  - on/off
iommu_platform=<bool>  - on/off
ioeventfd=<bool>       - on/off
any_layout=<bool>      - on/off
failover_pair_id=<str>
ats=<bool>             - on/off
x-pcie-lnksta-dllla=<bool> - on/off
x-disable-pcie=<bool>  - on/off
x-pcie-flr-init=<bool> - on/off
addr=<int32>           - Slot and optional function number, example: 06.0 or 06
x-pcie-extcap-init=<bool> - on/off
x-pcie-lnkctl-init=<bool> - on/off
romfile=<str>
rombar=<uint32>
event_idx=<bool>       - on/off
disable-modern=<bool>
class=<uint32>
migrate-extra=<bool>   - on/off
virtio-pci-bus-master-bug-migration=<bool> - on/off
modern-pio-notify=<bool> - on/off
indirect_desc=<bool>   - on/off
command_serr_enable=<bool> - on/off
```

We're not implementing all of them, but you can see it does support the
multifunction parameter, let's move that up into the common path.

Reply:
Sorry yeah completely missed this. Was able to check using `qemu-system-aarch64 --device virtio-serial,help` for aarch64